TAP-2.md

```diff
@@ -37,123 +37,84 @@ Provide all on-chain liquidity in a familiar limit orderbook interface to read (
+```rust
+/// Generic router interface for finding and executing trades
+pub trait Router {
+    fn get_amount_out(
+        &self,
+        amount_in: U256,
+        token_in: &Address,
+        token_out: &Address,
+    ) -> Result<GetAmountOutResult, Error>;
 
-// Operation Results
-type SizeQuote = {
-  effectivePrice: string
-  route: Route
-  gasCost: Level['gasCost']
-  priceImpact: string
+    fn encode_solution(
+        &self,
+        result: GetAmountOutResult,
+        sender: Address,
+        receiver: Address,
+    ) -> Result<RoutingSolution, Error>;
 }
-
-type DepthAtPrice = {
-  totalSize: string
-  averagePrice: string
-  route: Route
-  gasCost: Level['gasCost']
+```
+### 3. Price Level Calculator 
+Calculates discrete price levels for the orderbook using router simulations.
+```rust
+pub struct PriceLevel {
+    pub price: U256,
+    pub size: U256,
+    pub solution: RoutingSolution,
 }
 
-// Event System
-type OrderbookEvent = {
-  type: 'price_level_changed'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
-  side: Side
-  price: string
-  newSize: string
-} | {
-  type: 'book_reset'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
+pub struct PriceLevelCalculator {
+    router: Box<dyn Router>,
+    price_points: u32,
 }
-
-// Core Operations
-type Operations = {
-  getOrderbook: (pair: TokenPair, options?: OrderbookOptions) => Promise<OrderbookState>
-  
-  getQuoteForSize: (
-    pair: TokenPair,
-    size: string,
-    side: Side,
-    options?: OrderbookOptions
-  ) => Promise<SizeQuote>
-  
-  getSizeAtLimit: (
-    pair: TokenPair,
-    limitPrice: string,
-    side: Side,
-    options?: OrderbookOptions
-  ) => Promise<DepthAtPrice>
-  
-  findOptimalSize: (
-    pair: TokenPair,
-    targetPrice: string,
-    options?: FinetuneOptions
-  ) => Promise<SizeQuote>
-  
-  subscribeToEvents: (
+```
+### 4. Orderbook Manager
+Maintains orderbook state and handles updates as new protocol states arrive.
+```rust
+pub struct OrderbookState {
     pair: TokenPair,
-    onEvent: (event: OrderbookEvent) => void
-  ) => { unsubscribe: () => void }
-  
-  getSupportedTokens: () => Promise<string[]>
-  getSupportedPairs: () => Promise<TokenPair[]>
+    timestamp: u64,
+    block_number: u64,
+    bids: Vec<PriceLevel>,
+    asks: Vec<PriceLevel>,
+}
+pub struct OrderbookManager {
+    calculator: PriceLevelCalculator,
+    current_books: HashMap<TokenPair, OrderbookState>,
 }
 ```
+## Integration Flow
+1. Protocol Stream Manager connects to Tycho Indexer and receives real-time updates.
+2. Router maintains a cache of which pools are used in any solution for each orderbook.
+3. When pool states update:
+   - Router checks if any updated pools are used in any solution for a given orderbook.
+   - If yes, the entire orderbook needs to be recalculated as the optimal routes and prices at every amount level may have changed.
+4. For affected orderbooks:
+   - Every amount level must be recalculated.
```

Comment:
Not just the current level and any higher levels? I guess it depends on how you calculate the levels...

Reply:
Lower levels might also have changed, if a pool now has a better price.